include/sst/effects/Bonsai.h

```diff
@@ -0,0 +1,1202 @@
+
+        switch ((b_params)idx)
+        {
+        case b_bass_boost:
+            return result.withName("Amount").withRange(0.f, 1.f).withDefault(0.25f);
+        case b_bass_distort:
+            return result.withName("Distort").withRange(0.f, 3.f).withDefault(1.f);
+        case b_tape_sat:
+            return result.withName("Tape Sat").withRange(0.f, 1.f).withDefault(0.25f);
+        case b_tape_dist_mode:
+            return result.withType(pmd::INT)
+                .withName("Tape Dist Mode")
+                .withRange(bdm_inv_sinh, bdm_sine)
+                .withDefault(bdm_tanh)
+                .withUnorderedMapFormatting({{bdm_inv_sinh, "Inverse Sinh"},
+                                             {bdm_tanh, "Tanh"},
+                                             {bdm_tanh_approx_foldback, "Foldback"},
+                                             {bdm_sine, "Sine-Tanh Combo"}});
+        case b_noise_sensitivity:
+            return result.withName("Sensitivity").withRange(0.f, 1.f).withDefault(0.25f);
+        case b_noise_gain:
+            return result.withName("Gain").asDecibelNarrow().withDefault(0.f);
+        case b_dull:
+            return result.withName("Dull").withRange(0.f, 1.f).withDefault(0.f);
+        case b_gain_in:
+            return result.withName("Input Gain").asDecibelNarrow().withDefault(0.f);
+        case b_gain_out:
+            return result.withName("Output Gain").asDecibelNarrow().withDefault(0.f);
+        case b_mix:
+            return result.withName("Mix").asPercent().withDefault(1.f);
+        case b_num_params:
+            throw std::logic_error("getParam called with num_params");
+        }
+        return result;
+    }
+
+    int ringout_value = -1;
+
+    float last[89] = {};
+    float sr_div2pi = Bonsai<FXConfig>::sampleRate() / (2 * M_PI);
+    const float coef0 = freq_sr2pi_to_alpha(0, sr_div2pi);
+    const float coef10 = freq_sr2pi_to_alpha(10, sr_div2pi);
+    const float coef20 = freq_sr2pi_to_alpha(20, sr_div2pi);
+    const float coef30 = freq_sr2pi_to_alpha(20, sr_div2pi);
+    const float coef50 = freq_sr2pi_to_alpha(50, sr_div2pi);
+    const float coef99 = freq_sr2pi_to_alpha(99, sr_div2pi);
+    const float coef100 = freq_sr2pi_to_alpha(100, sr_div2pi);
+    const float coef160 = freq_sr2pi_to_alpha(160, sr_div2pi);
+    const float coef200 = freq_sr2pi_to_alpha(200, sr_div2pi);
+    const float coef500 = freq_sr2pi_to_alpha(500, sr_div2pi);
+    const float coef700 = freq_sr2pi_to_alpha(700, sr_div2pi);
+    const float coef1000 = freq_sr2pi_to_alpha(1000, sr_div2pi);
+    const float coef1200 = freq_sr2pi_to_alpha(1200, sr_div2pi);
+    const float coef1280 = freq_sr2pi_to_alpha(1280, sr_div2pi);
+    const float coef2000 = freq_sr2pi_to_alpha(2000, sr_div2pi);
+    const float coef3000 = freq_sr2pi_to_alpha(3000, sr_div2pi);
+    const float coef4690 = freq_sr2pi_to_alpha(4690, sr_div2pi);
+    const float coef8000 = freq_sr2pi_to_alpha(8000, sr_div2pi);
```

Comment:
right so we need to think hard about what happens when sample rates change. I should add an event in the base class for this which will call you and let you reset. Unfortunately oversampling in DAWs can reset sample rates on running effects. We can merge a first version before we fix this but we should open an issue if we do, while I try to figure out how to do an on-sample-change notification into the effects framework

Reply:
yeah i wondered that at one point when i read about airwindows needing some call to reset stuff